projects/portal/src/app/pages/delegate/validators/validators.usecase.service.ts

```diff
@@ -0,0 +1,166 @@
+        }, 0),
+      ),
+    );
+    this.validatorsWithShare$ = this.allValidatorsTokens$.pipe(
+      withLatestFrom(this.validatorsList$),
+      map(([allTokens, validators]) => {
+        if (!allTokens) {
+          return [];
+        }
+        // calculate validator share
+        const validatorsWithShare = validators?.map((validator) => {
+          const val = validator;
+          const share = Number(validator.tokens) / allTokens;
+          return { val, share };
+        });
+        // sort by share
+        const validatorsWithSort = validatorsWithShare?.sort((x, y) => y.share - x.share);
+
+        return validatorsWithSort || [];
+      }),
+    );
+
+    this.accAddress$ = this.walletService.currentStoredWallet$.pipe(
+      filter((wallet): wallet is StoredWallet => wallet !== undefined && wallet !== null),
+      map((wallet) => cosmosclient.AccAddress.fromString(wallet.address)),
+    );
+
+    this.delegationsInService$ = this.accAddress$.pipe(
+      mergeMap((address) => this.cosmosRest.getDelegatorDelegations$(address)),
+      map((result) => (result?.delegation_responses ? result.delegation_responses : undefined)),
+    );
+    this.totalDelegation$ = this.delegationsInService$.pipe(
+      map((delegations) =>
+        delegations?.reduce(
+          (sum, element) =>
+            element.balance?.denom == 'uguu' ? sum + Number(element.balance?.amount) : sum,
+          0,
+        ),
+      ),
+    );
+    this.delegatedValidatorsInService$ = combineLatest([
+      this.validatorsList$,
+      this.delegationsInService$,
+    ]).pipe(
+      map(([validators, delegations]) =>
+        delegations?.map((delegation) =>
+          validators?.find(
+            (validator) => validator.operator_address == delegation.delegation?.validator_address,
+          ),
+        ),
+      ),
+    );
+
+    this.unbondingDelegations$ = this.delegatedValidatorsInService$.pipe(
+      withLatestFrom(this.accAddress$),
+      concatMap(([validators, accAddress]) => {
+        if (!validators) {
+          console.log('0');
+          return [];
+        }
+
+        const valAddressList: (cosmosclient.ValAddress | undefined)[] = validators.map(
+          (validator) => {
+            if (!validator?.operator_address) {
+              console.log('1');
+              return undefined;
+            }
+            try {
+              return cosmosclient.ValAddress.fromString(validator?.operator_address);
+            } catch (error) {
+              console.error(error);
+              return undefined;
+            }
+          },
+        );
+        if (!valAddressList) return [];
+
+        const unbondingDelegationList = Promise.all(
+          valAddressList.map((valAddress) => {
+            if (!valAddress) return undefined;
+            return this.cosmosRest.getUnbondingDelegation$(valAddress, accAddress).toPromise();
+          }),
+        );
+        return unbondingDelegationList;
+      }),
+    );
```

Comment:
As a guideline, try reducing the contents initialized in the constructor.

Reply:
Thank you for your review. Removed all code in the constructor.